### Data Wrangling

* __Data wrangling__ is the process of converting data from the initial format to a format that may be better for analysis.

#### Import Pandas

You are running the lab in your browser, so we will install the libraries using piplite.

__import piplite__

__await piplite.install(['pandas'])__

__await piplite.install(['matplotlib'])__

Install specific version of libraries used in lab:

!mamba install pandas==1.3.3

!mamba install numpy=1.21.2

In [1]:
import pandas as pd
import matplotlib.pylab as plt

#### Reading the dataset from the URL and adding the related headers

First, we assign the URL of the dataset to "filename".

In [2]:
filename="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DA0101EN-SkillsNetwork/labs/Data%20files/auto.csv"

In [3]:
# We create a Python list headers containing name of headers.

headers= ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

Use the Pandas method __read_csv()__ to load the data from the web address. Set the parameter "names" equal to the Python list "headers".

In [4]:
df=pd.read_csv(filename, names=headers)

Use the method __head()__ to display the first five rows of the dataframe.

In [5]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

So, how do we identify all those missing values and deal with them?

#### How to work with missing data?

Steps for working with missing data:

1.Identify missing data

2.Deal with missing data

3.Correct data format

### Identify and handle missing values

#### Identify missing values

#### Convert "?" to NaN

In the car dataset, missing data comes with the question mark "?". We replace "?" with __NaN (Not a Number),__ Python's default missing value marker for reasons of computational speed and convenience. Here we use the function:

.replace(A, B, inplace = True) 

to replace A by B.

In [6]:
import numpy as np

#replace "?" to NaN

df.replace("?", np.NaN, inplace=True)

In [7]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

#### Evaluating for Missing Data

The missing values are converted by default. We use the following functions to identify these missing values. There are two methods to detect missing data:

1. **.isnull()**
2. **.notnull()**

The output is a boolean value indicating whether the value that is passed into the argument is in fact missing data.

In [8]:
missing_data=df.isnull()
missing_data.head(5)

symboling  normalized-losses   make  fuel-type  aspiration  num-of-doors  \
0      False               True  False      False       False         False   
1      False               True  False      False       False         False   
2      False               True  False      False       False         False   
3      False              False  False      False       False         False   
4      False              False  False      False       False         False   

   body-style  drive-wheels  engine-location  wheel-base  ...  engine-size  \
0       False         False            False       False  ...        False   
1       False         False            False       False  ...        False   
2       False         False            False       False  ...        False   
3       False         False            False       False  ...        False   
4       False         False            False       False  ...        False   

   fuel-system   bore  stroke  compression-ratio  horsepower  peak-rpm  \
0        False  False   False              False       False     False   
1        False  False   False              False       False     False   
2        False  False   False              False       False     False   
3        False  False   False              False       False     False   
4        False  False   False              False       False     False   

   city-mpg  highway-mpg  price  
0     False        False  False  
1     False        False  False  
2     False        False  False  
3     False        False  False  
4     False        False  False  

[5 rows x 26 columns]

__"True"__ means the value is a missing value while __"False"__ means the value is not a missing value.

#### Count missing values in each column

Using a for loop in Python, we can quickly figure out the number of missing values in each column. As mentioned above, "True" represents a missing value and "False" means the value is present in the dataset. In the body of the for loop the method __".value_counts()"__ counts the number of "True" values.

In [9]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print("")

symboling
False    205
Name: symboling, dtype: int64

normalized-losses
False    164
True      41
Name: normalized-losses, dtype: int64

make
False    205
Name: make, dtype: int64

fuel-type
False    205
Name: fuel-type, dtype: int64

aspiration
False    205
Name: aspiration, dtype: int64

num-of-doors
False    203
True       2
Name: num-of-doors, dtype: int64

body-style
False    205
Name: body-style, dtype: int64

drive-wheels
False    205
Name: drive-wheels, dtype: int64

engine-location
False    205
Name: engine-location, dtype: int64

wheel-base
False    205
Name: wheel-base, dtype: int64

length
False    205
Name: length, dtype: int64

width
False    205
Name: width, dtype: int64

height
False    205
Name: height, dtype: int64

curb-weight
False    205
Name: curb-weight, dtype: int64

engine-type
False    205
Name: engine-type, dtype: int64

num-of-cylinders
False    205
Name: num-of-cylinders, dtype: int64

engine-size
False    205
Name: engine-size, dtype: int64

fuel-system
Fa

Based on the summary above, each column has 205 rows of data and seven of the columns containing missing data:

1. "normalized-losses": 41 missing data
2. "num-of-doors": 2 missing data
3. "bore": 4 missing data
4. "stroke" : 4 missing data
5. "horsepower": 2 missing data
6. "peak-rpm": 2 missing data
7. "price": 4 missing data

In [10]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               NaN  alfa-romero       gas        std          two   
1          3               NaN  alfa-romero       gas        std          two   
2          1               NaN  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [11]:
save_Data=df.to_csv("Wrangling1.csv", index= False)

### Deal with missing data

#### How to deal with missing data?

##### 1. Drop data

    a. Drop the whole row

    b. Drop the whole column

##### 2. Replace data

    a. Replace it by mean

    b. Replace it by frequency

    c. Replace it based on other functions

Whole columns should be dropped only if most entries in the column are empty. In our dataset, none of the columns are empty enough to drop entirely. We have some freedom in choosing which method to replace data; however, some methods may seem more reasonable than others. We will apply each method to many different columns:

#### Replace by mean:

* "normalized-losses": 41 missing data, replace them with mean
* "stroke": 4 missing data, replace them with mean
* "bore": 4 missing data, replace them with mean
* "horsepower": 2 missing data, replace them with mean
* "peak-rpm": 2 missing data, replace them with mean

* Replace by frequency:

    * "num-of-doors": 2 missing data, replace them with "four".
         * Reason: 84% sedans is four doors. Since four doors is most frequent, it is most likely to occur

#### Drop the whole row:

* "price": 4 missing data, simply delete the whole row

 * Reason: price is what we want to predict. Any data entry without price data cannot be used for prediction; therefore any row now without price data is not useful to us

#### Calculate the mean value for the "normalized-losses" column

In [12]:
avg_norm_loss= df["normalized-losses"].astype("float").mean(axis=0)
print("Average Normalized Losses : ",avg_norm_loss)                                              

Average Normalized Losses :  122.0


#### Replace "NaN" with mean value in "normalized-losses" column

In [13]:
df["normalized-losses"].replace(np.NaN, avg_norm_loss, inplace=True)

#### Calculate the mean value for the "bore" column

In [14]:
avg_bore=df["bore"].astype("float").mean(axis=0)
print("Average of bore:", avg_bore)

Average of bore: 3.3297512437810957


#### Replace "NaN" with the mean value in the "bore" column

In [15]:
df["bore"].replace(np.NaN, avg_bore, inplace=True)

#### Calculate the mean value for the "Stoke" column

In [16]:
avg_stoke=df["stroke"].astype("float").mean(axis=0)
print("Average of Stoke :", avg_stoke)

Average of Stoke : 3.2554228855721337


#### Replace NaN in "stroke" column with the mean value.

In [17]:
df["stroke"].replace(np.NaN, avg_stoke, inplace=True)

#### Calculate the mean value for the "horsepower" column

In [18]:
avg_horsepower = df['horsepower'].astype('float').mean(axis=0)
print("Average horsepower:", avg_horsepower)

Average horsepower: 104.25615763546799


#### Replace "NaN" with the mean value in the "horsepower" column

In [19]:
df['horsepower'].replace(np.nan, avg_horsepower, inplace=True)

#### Calculate the mean value for "peak-rpm" column

In [20]:
avg_peakrpm=df['peak-rpm'].astype('float').mean(axis=0)
print("Average peak rpm:", avg_peakrpm)

Average peak rpm: 5125.369458128079


#### Replace "NaN" with the mean value in the "peak-rpm" column

In [21]:
df['peak-rpm'].replace(np.nan, avg_peakrpm, inplace=True)

To see which values are present in a particular column, we can use the ".value_counts()" method:

In [22]:
df["num-of-doors"].value_counts()

four    114
two      89
Name: num-of-doors, dtype: int64

We can see that four doors are the most common type. We can also use the ".idxmax()" method to calculate the most common type automatically:

In [23]:
df["num-of-doors"].value_counts().idxmax()

'four'

In [24]:
df["num-of-doors"].value_counts().idxmin()

'two'

The replacement procedure is very similar to what we have seen previously:

In [25]:
#replace the missing 'num-of-doors' values by the most frequent 

df["num-of-doors"].replace(np.NaN, "four", inplace=True)

Finally, let's drop all rows that do not have price data:

In [26]:
# simply drop whole row with NaN in "price" column
df.dropna(subset=["price"], axis=0, inplace=True)

# reset index, because we droped two rows
df.reset_index(drop=True, inplace=True)

In [27]:
df.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3             122.0  alfa-romero       gas        std          two   
1          3             122.0  alfa-romero       gas        std          two   
2          1             122.0  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [28]:
df.tail()

symboling normalized-losses   make fuel-type aspiration num-of-doors  \
196         -1                95  volvo       gas        std         four   
197         -1                95  volvo       gas      turbo         four   
198         -1                95  volvo       gas        std         four   
199         -1                95  volvo    diesel      turbo         four   
200         -1                95  volvo       gas      turbo         four   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
196      sedan          rwd           front       109.1  ...          141   
197      sedan          rwd           front       109.1  ...          141   
198      sedan          rwd           front       109.1  ...          173   
199      sedan          rwd           front       109.1  ...          145   
200      sedan          rwd           front       109.1  ...          141   

     fuel-system  bore  stroke compression-ratio horsepower  peak-rpm  \
196         mpfi  3.78    3.15               9.5        114      5400   
197         mpfi  3.78    3.15               8.7        160      5300   
198         mpfi  3.58    2.87               8.8        134      5500   
199          idi  3.01    3.40              23.0        106      4800   
200         mpfi  3.78    3.15               9.5        114      5400   

    city-mpg highway-mpg  price  
196       23          28  16845  
197       19          25  19045  
198       18          23  21485  
199       26          27  22470  
200       19          25  22625  

[5 rows x 26 columns]

__Good!__ Now, we have a dataset with no missing values.

In [29]:
df.to_csv("Data Wrangling2.csv", index=False)